RHDEVDOCS-5791: Documentation: Shared resource CSI driver GA (for 1.1) (#83508)
openshift/openshift-docs: modules/builds-running-entitled-builds-with-sharedsecret-objects.adoc

:_mod-docs-content-type: PROCEDURE
[id="builds-running-entitled-builds-with-sharedsecret-objects_{context}"]
= Running builds using SharedSecret objects

You can use a `SharedSecret` object to securely access the entitlement keys of a cluster in builds.

The `SharedSecret` object allows you to share and synchronize secrets across namespaces.

[IMPORTANT]
====

> The Shared Resource CSI Driver feature is now generally available in link:https://docs.redhat.com/en/documentation/builds_for_red_hat_openshift/1.1[{builds-v2title} 1.1]. This feature is now deprecated in {product-title}. To use this feature, ensure you are using {builds-v2title} 1.1 or a more recent version.

Reviewer: Agree on dropping the bit that states the history of the CSI Driver and only focus on the GA statement.

** Determine if the `builder` service account available to you in your namespace is allowed to use the given `SharedSecret` CR instance. In other words, you can run `oc adm policy who-can use <identifier_of_specific_SharedSecret>` to see if the `builder` service account in your namespace is listed.

[NOTE]
====
If neither of the last two prerequisites in this list are met, establish, or ask someone to establish, the necessary role-based access control (RBAC) so that you can discover `SharedSecret` CR instances and enable service accounts to use `SharedSecret` CR instances.
====

.Procedure

. Create a `SharedSecret` object instance with the cluster's entitlement secret by running the `oc apply` command:
+
[IMPORTANT]
====
You must have cluster administrator permissions to create `SharedSecret` objects.
====
+
[source,terminal]
----
$ oc apply -f -<<EOF
apiVersion: sharedresource.openshift.io/v1alpha1
kind: SharedSecret
metadata:
  name: share-rhel-entitlement
spec:
  secretRef:
    name: etc-pki-entitlement
    namespace: openshift-config-managed
EOF
----

> . Create a role to grant permission to access the `SharedSecret` object and a `RoleBinding` object that grants permission to access the `SharedSecret` object by running the following command:

Reviewer 2: ```suggestion
. Create a `ClusterRole` to grant permission to access the `SharedSecret` object and a `RoleBinding` object that grants permission to access the `SharedSecret` object by running the following command:
```

+
[source,terminal]
----
$ oc new-project buildah-entitlement
$ oc apply -f -<<EOF
---
apiVersion: rbac.authorization.k8s.io/v1

> kind: Role

Reviewer 2: This needs to be `ClusterRole` because `SharedSecret` is cluster-scoped. See https://kubernetes.io/docs/reference/access-authn-authz/rbac/#role-and-clusterrole

```suggestion
kind: ClusterRole
```

@apoorvajagtap @sayan-biswas do we have a test script to verify?

metadata:
  name: use-shared-secret
rules:
  - apiGroups:
      - sharedresource.openshift.io
    resources:
      - sharedsecrets
    resourceNames:
      - share-entitlement-secret
    verbs:
      - use
---
apiVersion: rbac.authorization.k8s.io/v1
kind: RoleBinding
metadata:
  name: use-shared-secret
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: Role
  name: use-shared-secret
subjects:
  - kind: ServiceAccount
    name: pipeline

>     namespace: buildah-entitlement

Reviewer 2: Remove namespace here - this needs to reference the `pipeline` Service Account in the namespace where the build occurs.

```suggestion

```

EOF

> ----

Reviewer 2: We are missing a step from [BUILD-965](https://issues.redhat.com/browse/BUILD-965) - to get us to GA, we reduced the default set of permissions granted to the CSI driver's service account. Admins need to explicitly grant the driver permission to access a given shared resource.

The procedure described in https://github.com/openshift/csi-driver-shared-resource is a bit too permissive, particularly for this use case. We can do better, and I'll try and post an example.

. Create a build with relevant mount details using `buildah` buildStrategy:
+
[source,terminal]
----
$ oc apply -f -<<EOF
apiVersion: shipwright.io/v1beta1
kind: Build
metadata:
  name: buildah-rhel-entitlement-build
spec:
  source:
    type: Git
    git:
      url: <location of operator's organization>
      revision: quick-tests
    contextDir: rh-builds
  strategy:
    name: buildah-csi
    kind: ClusterBuildStrategy

>   paramValues:

Reviewer 2: @apoorvajagtap I believe we now have a GitHub repository for sample code, and the buildah task has/will be updated to have the appropriate volume defined?

Reviewer: Agree that we should add a Git repo in there that contains the example

Reviewer 3: Yes, I informed Shivani about this during our sync-up. She'll be adding the correct reference.
Also, @shivanisathe25 for 1.2 the `strategy.name` should be set to `buildah`. (for 1.1, with the additional step 1, the strategy.name is correct: `buildah-csi`)

```suggestion
    git:
      url: https://github.com/redhat-openshift-builds/samples/
    contextDir: buildah-build
  strategy:
    name: buildah-csi
    kind: ClusterBuildStrategy
  paramValues:
```

  - name: dockerfile
    value: Dockerfile
  volumes:

>   - csi:
>       driver: csi.sharedresource.openshift.io
>       readOnly: true
>       volumeAttributes:
>         sharedSecret: share-entitlement-secret
>     name: rhel-entitlement

Reviewer 2: Names need to match:

- Volume name must match what we are defining in the cluster build strategy: https://github.com/redhat-openshift-builds/operator/blob/main/config/shipwright/build/strategy/buildah.yaml#L206-L208
- `sharedSecret` must match the name of the `SharedSecret` object above.

```suggestion
  - csi:
      driver: csi.sharedresource.openshift.io
      readOnly: true
      volumeAttributes:
        sharedSecret: share-rhel-entitlement
    name: etc-pki-entitlement
```

  - name: buildah-auth
    secret:

>       secretName: buildah-auth

Reviewer 2: This volume does not exist - please remove.

  output:

>     image: image-registry.openshift-image-registry.svc:5000/buildah-entitlement/sample-go-app

Reviewer 2: We might need a note here - for things that push to the internal registry, the URL can change depending on the namespace the build runs in. This is a persistent challenge with Shipwright vs. BuildConfig.

EOF
----

> . Create the buildRun referencing the rhel-entitlement-build:
> +
> [source,terminal]
> ----
> $ oc apply -f -<<EOF
> apiVersion: shipwright.io/v1beta1
> kind: BuildRun
> metadata:
>   name: buildah-rhel-entitlement-buildrun
> spec:
>   build:
>     name: buildah-rhel-entitlement-build
> EOF
> ----

Reviewer 2: @siamaksade should we instruct developers to use the `shp` command line instead?

```suggestion
. Run a build for `buildah-rhel-entitlement-build`:
+
[source,terminal]
----
$ shp build run buildah-rhel-entitlement-build
----
```

Author: Adding about the `shp` command as note

[NOTE]
====
* You must include the command to remove the `/etc/rhsm-host` directory and all its contents in the Dockerfile before executing any `yum` or `dnf` commands.
* Use the link:https://access.redhat.com/downloads/content/package-browser[Red Hat Package Browser] to find the correct repositories for your installed packages.

> * You must restore the `/etc/rhsm-host` symbolic link to keep your image compatible with other Red Hat container images.

Reviewer 2: Where is the original `Dockerfile` that this footnote was pointing to?

* You must set `readOnly` to `true` to mount the shared resource in the build.

> * Reference the name of the `SharedSecret` object to include it in the build. 

Reviewer 2: These likewise need to be appropriately footnoted, pointing to an appropriate location in the YAML file. ASCIIDoc supports this.

:_mod-docs-content-type: PROCEDURE
[id="builds-running-entitled-builds-with-sharedsecret-objects_{context}"]
= Running builds using SharedSecret objects

You can use a `SharedSecret` object to securely access the entitlement keys of a cluster in builds.

The `SharedSecret` object allows you to share and synchronize secrets across namespaces.

[IMPORTANT]
====
The Shared Resource CSI Driver feature is now generally available in link:https://docs.redhat.com/en/documentation/builds_for_red_hat_openshift/1.1[{builds-v2title} 1.1]. This feature is now deprecated in {product-title}. To use this feature, ensure you are using {builds-v2title} 1.1 or a more recent version.
====

.Prerequisites

> * You have enabled the `TechPreviewNoUpgrade` feature set by using the feature gates. For more information, see xref:../../nodes/clusters/nodes-cluster-enabling-features.adoc#nodes-cluster-enabling[Enabling features using feature gates].

Reviewer 3: Do we still need this? As we are announcing the feature as GA, and it's available with builds, I don't think there's any need to enable `TechPreviewNoUpgrade` now.

* You must have permission to perform the following actions:
** Create build configs and start builds.
** Discover which `SharedSecret` custom resource (CR) instances are available by entering the `oc get sharedsecrets` command and getting a non-empty list back.

> ** Determine if the `builder` service account available to you in your namespace is allowed to use the given `SharedSecret` CR instance. In other words, you can run `oc adm policy who-can use <identifier_of_specific_SharedSecret>` to see if the `builder` service account in your namespace is listed.

Reviewer 3: @sayan-biswas @adambkaplan is there any specific dependency on `builder` SA now (with builds operator)?

Reviewer 2: No specific dependency on the `builder` service account. This can apply to any service account.

```suggestion
** Determine if any service account in your namespace is allowed to use the given `SharedSecret` instance. In other words, you can run `oc adm policy who-can use <identifier_of_specific_SharedSecret>` to see if the desired service account in your namespace is listed.
```
